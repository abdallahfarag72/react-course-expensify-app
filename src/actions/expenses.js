import { v4 as uuidv4 } from 'uuid'
import database from "../firebase/firebase";
import { getDatabase, ref, set, remove, update, onValue, off, push, onChildRemoved, onChildChanged, onChildAdded, get } from "firebase/database";


// -------Action Generators for Expenses-------

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData
        
        const expense = { description, note, amount, createdAt }

        return push(ref(database, 'expenses'), expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return remove(ref(database, `expenses/${id}`)).then(() => {
            dispatch(removeExpense({ id }))
        })
    }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return (dispatch) => {
        const expenses = []
        return get(ref(database, 'expenses')).then((snapshot) => {
            snapshot.forEach(childSnapshot => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            });
            dispatch(setExpenses(expenses))
        })
    }
}