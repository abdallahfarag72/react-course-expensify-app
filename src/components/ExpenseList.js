import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";


// exporting unconnected version (unconnected from the store) for testing only 
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            )   :   (
                props.expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
            )
        }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

// exporting the connected version to be used in the app
export default connect(mapStateToProps)(ExpenseList)

