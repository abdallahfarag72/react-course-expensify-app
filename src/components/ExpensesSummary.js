import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import getVisibleExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";
import numeral from "numeral";

export const ExpensesSummary = (props) => {
    const s = props.expensesCount === 1 ? '' : 's'
    return (
        <div className="page-header">
         <div className="content-container">
          <h1 className="page-header__title">Viewing <span>{props.expensesCount}</span> expense{s} totalling <span>{numeral(props.expensesTotal / 100).format('$0,0.00')}</span></h1>
          <div className="page-header__actions" >
           <Link className="button" to='/create' >Add Expense</Link>
          </div>
         </div>
        </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)

