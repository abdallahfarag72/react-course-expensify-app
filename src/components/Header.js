import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to='/' className={({ isActive }) => isActive ? 'is-active' : undefined}>Home</NavLink>
        <NavLink to='/create' className={({ isActive }) => isActive ? 'is-active' : undefined}>create Expense</NavLink>
    </header>
)

export default Header
