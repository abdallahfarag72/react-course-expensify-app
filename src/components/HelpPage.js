import React from "react";

const HelpPage = () => (
    <div>
        this is my help page
    </div>
)

export default HelpPage