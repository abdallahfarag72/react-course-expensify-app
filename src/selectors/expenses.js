import moment from "moment"
// -------Get Visible Expenses---------

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter(expense => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true //if startDate is undefined, don't filter and return all expenses. if its a number, filter accordingly 
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt > b.createdAt ? -1 : 1 //sorting by most recent
        } else if (sortBy === 'amount') {
            return a.amount > b.amount ? -1 : 1 //sorting by highest amount
        }
    })
}