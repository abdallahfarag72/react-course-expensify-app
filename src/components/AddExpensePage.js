import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import { addExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.navigate('/')
    }

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}


const mapDispatchToProp = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

const WithNavigate = (props) => {
    const navigate = useNavigate()
    return <AddExpensePage {...props} navigate={navigate} />
}

export default connect(undefined, mapDispatchToProp)(WithNavigate)