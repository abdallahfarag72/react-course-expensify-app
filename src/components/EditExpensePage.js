import React from "react";
import { useParams } from 'react-router-dom';
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, startRemoveExpense } from "../actions/expenses";
import { useNavigate } from "react-router-dom";

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.id, expense)
        this.props.navigate('/')
    }
    onClick = () => {
        this.props.startRemoveExpense({ id: this.props.id })
        this.props.navigate('/')
    }

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onClick}>Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    // let { id } = useParams();
    // return {
    //     expense: state.expenses.find(expense => {
    //         return expense.id === id 
    //     }),

    // }
    
    
    // for some reason this code solved the hooks error
    const params = { id: window.location.pathname.split("/")[2] }
    return {
        expense: state.expenses.find((expense) => expense.id === params.id)
    }
}

const WithNavigateAndParams = (props) => {
    const { id } = useParams()
    const navigate = useNavigate()
    return <EditExpensePage {...props} navigate={navigate} id={id} />
}

const mapDispatchToProp = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: ({ id }) => dispatch(startRemoveExpense({ id }))
})

export default connect(mapStateToProps, mapDispatchToProp)(WithNavigateAndParams)