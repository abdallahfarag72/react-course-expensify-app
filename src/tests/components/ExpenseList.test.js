import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { ExpenseList } from "../../components/ExpenseList";

test('Should render expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
})

test('should render expene list without expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})