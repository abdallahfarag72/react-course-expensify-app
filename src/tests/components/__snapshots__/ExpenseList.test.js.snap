// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={19542}
    createdAt={0}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={230}
    createdAt={345600000}
    description="Coffee"
    id="3"
    key="3"
    note=""
  />
  <ExpenseListItem
    amount={1234568}
    createdAt={-345600000}
    description="Gum"
    id="4"
    key="4"
    note=""
  />
</div>
`;

exports[`should render expene list without expenses 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
